Sources/KituraContracts/Contracts.swift

```diff
@@ -59,15 +94,144 @@ public struct RequestError: RawRepresentable, Equatable, Hashable, Comparable, E
     /// A human-readable description of the error code.
     public let reason: String
 
+    /**
+     Representation of the error body-an object representing further
+     details of the failure.
+
+     The value may be:
+     - `nil` if there is no body
+     - a (type-erased) Codable object if the error was initialized with `init(_:body:)`
+     - bytes of data and a signifier of the format in which they are stored (eg: JSON)
+       if the error was initialized with `init(_:bodyData:format:)`
+
+     ### Usage example: ###
+     ````
+     if let errorBody = error.body {
+         switch error.body {
+            case let .codable(body): ... // body is Codable
+            case let .data(bytes, format): ... // bytes is Data, format is BodyFormat
+         }
+     }
+     ````
+
+     - Note: If you need a Codable representation and the body is data, you
+             can call the `bodyAs(_:)` function to get the converted value
+     */
+    public private(set) var body: ErrorBody? = nil
+
+    // A closure used to hide the generic type of the Codable body
+    // for later encoding to Data
+    private var bodyDataEncoder: ((BodyFormat) throws -> Data)? = nil
+
+    /**
+     Returns the error body encoded into bytes in a given format (eg: JSON).
+
+     This function should be used if the RequestError was created using
+     `init(_:body:)`, otherwise it will return `nil`.
+
+     - Note: This function is primarily intended for use by the Kitura Router so
+             that it can encode and send a custom error body returned from
+             a codable route.
+
+     ### Usage Example: ###
+     ````
+     do {
+         if let errorBodyData = try error.encodeBody(.json) {
+             ...
+         }
+     } catch {
+         // Handle the failure to encode
+     }
+     ````
+     - parameter `BodyFormat` describes the format that should be used
+                 (for example: `BodyFormat.json`)
+     - returns the `Data` object or `nil` if there is no body, or if the
+               error was not initialized with `init(_:body:)`
+     - throws an `EncodingError` if the encoding fails
+     - throws an `UnsupportedBodyFormatError` if the provided `BodyFormat`
+              is not supported
+     */
+    public func encodeBody(_ format: BodyFormat) throws -> Data? {
+        guard case .codable? = body else { return nil }
```

Comment:
I'm confused. Doesn't this return nil if we're not `.codable`? Don't we want to return nil if we _are_ `.codable`?

Reply:
As discussed in person, this function is for encoding the Codable body into Data, so the "source" of the body must be Codable rather than Data.
I will change the doc to "Returns the codable error body..." in the hope this clarifies the behaviour.

Side-note: I did consider a different API where the encode/decode was done transparently (ie you could just access the body as Codable or Data and it would work regardless of how it was initialized, but it didn't turn out well so I went for this less magic API instead).